packages/x-utils/src/types/object.types.ts

```diff
@@ -0,0 +1,22 @@
+/**
+ * Tree-shakes the provided object returning another one with only the keys that match the given
+ * type.
+ *
+ * @remarks By default, ExcludeOptional is false, so optional properties are included.
+ *
+ * @param SomeObject - The object to tree-shake.
+ * @param Type - The type of the properties to preserve.
+ * @param ExcludeOptional - Flag to exclude optional properties.
+ *
+ * @public
+ */
+export type TreeShakeObject<SomeObject, Type, ExcludeOptional extends boolean = false> = {
```

Comment:
I used this name (`tree-shaking`) because it basically tree shakes an object, returning another one with the same shape but removing only those properties that don't match the passed type.

If you don't find it clear or sounds ambiguous, feel free to propose another one. We accept both cookies and better names.